source/common/conn_pool/conn_pool_base.cc

```diff
@@ -220,9 +218,12 @@ void ConnPoolImplBase::onStreamClosed(Envoy::ConnectionPool::ActiveClient& clien
   // increment as no capacity is freed up.
   // We don't update the capacity for HTTP/3 as the stream count should only
   // increase when a MAX_STREAMS frame is received.
-  if (trackStreamCapacity() && (client.remaining_streams_ > client.concurrent_stream_limit_ -
-                                                                client.numActiveStreams() - 1 ||
-                                had_negative_capacity)) {
+  // The capacity calculated by concurrency could be negative, in this case, effective client
+  // capacity was limited by concurrency, compare client.concurrent_stream_limit_ and
+  // client.numActiveStreams() directly to avoid overflow.
+  if (trackStreamCapacity() && ((client.concurrent_stream_limit_ < client.numActiveStreams() + 1) ||
+                                (client.remaining_streams_ > client.concurrent_stream_limit_ -
+                                                                 client.numActiveStreams() - 1))) {
```

Comment:
While I like the simplification of removing the negative capacity member variable, I think this if statement was already hard to read and is now really hard to read :-)

Let's break it up.  Something like 
// We don't update the capacity for HTTP/3
if (trackStreamCapacity()) {
  // If the effective client capacity was limited by max total streams...
  bool limited_by_concurrency = ...
  // The capacity calculated by concurrency could be negative [maybe explicitly call out SETTINGS frame
  bool has_newly_available_capacity = ...
 if (limited_by_concurrency && has_newly_available_capacity) {
   state.incr....
   }

Reply:
@alyssawilk updated!